File: БР1.2/Тимаков Егор/labs/src/main/java/org/rentservice/service/auth/JwtService.java
package org.rentservice.service.auth;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import org.springframework.stereotype.Service;

import java.util.Date;

@Service
public class JwtService {

    private static final String SECRET =
            "8PsTQh7t29YSlzijZMmFGSl1hZMeGUxgOTGghRNLv6B";

    public String generateToken(String email) {

        return Jwts.builder()
                .setSubject(email)
                .setIssuedAt(new Date())
                .setExpiration(
                        new Date(
                                System.currentTimeMillis()
                                        + 86400000
                        )
                )
                .signWith(
                        Keys.hmacShaKeyFor(
                                SECRET.getBytes()
                        ),
                        SignatureAlgorithm.HS256
                )
                .compact();
    }

    public String extractEmail(
            String token
    ) {

        return Jwts.parserBuilder()
                .setSigningKey(
                        Keys.hmacShaKeyFor(
                                SECRET.getBytes()
                        )
                )
                .build()
                .parseClaimsJws(token)
                .getBody()
                .getSubject();
    }

    public boolean validateToken(
            String token
    ) {

        try {

            extractEmail(token);

            return true;

        } catch (Exception e) {

            return false;
        }
    }

}

File: БР1.2/Цатинян Артём/labs/lab1/src/main/kotlin/ru/itmo/restaurantbooking/auth/service/JwtService.kt
package ru.itmo.restaurantbooking.auth.service

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Service
import ru.itmo.restaurantbooking.auth.config.JwtProperties
import ru.itmo.restaurantbooking.auth.domain.AuthenticatedUser
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.Date

@Service
class JwtService(
    private val jwtProperties: JwtProperties
) {

    private val key by lazy {
        Keys.hmacShaKeyFor(jwtProperties.secret.toByteArray(StandardCharsets.UTF_8))
    }

    fun generateToken(user: AuthenticatedUser): String {
        val now = Instant.now()

        return Jwts.builder()
            .subject(user.id.toString())
            .issuer(jwtProperties.issuer)
            .issuedAt(Date.from(now))
            .expiration(Date.from(now.plus(jwtProperties.accessTokenTtlMinutes, ChronoUnit.MINUTES)))
            .claim("email", user.email)
            .claim("role", user.role.name)
            .signWith(key)
            .compact()
    }

    fun parse(token: String): AuthenticatedUser {
        val claims = Jwts.parser().verifyWith(key).build().parseSignedClaims(token).payload
        return AuthenticatedUser(
            id = claims.subject.toLong(),
            email = claims["email", String::class.java],
            role = ru.itmo.restaurantbooking.user.domain.UserRole.valueOf(claims["role", String::class.java])
        )
    }
}
